### ISSUE TITLE
Setup Java does not ensure that `sbt` is installed on Ubuntu 24

### ISSUE THREAD

**Issue by jsoref** [actions/setup-java#712]
**Description:**
[Notice of breaking changes for GitHub Actions](https://github.blog/changelog/2024-11-05-notice-of-breaking-changes-for-github-actions/) warned that certain packages would not be present in Ubuntu 24.04, this included `sbt`.

This was tripped on by this repository and worked around by pinning backwards (from `ubuntu-latest` to `ubuntu-22.04`):
- #693

Unfortunately, anyone who legitimately expected "setup-java" to set up java for use with sbt which is listed as a supported feature of the actions/setup-java repository:
https://github.com/actions/setup-java/blob/7a6d8a8234af8eb26422e24e3006232cccaa061b/README.md?plain=1#L16
... doesn't get a working sbt environment.

**Task version:**
Specify the task version `actions/setup-java@v4`, as in:
https://github.com/GarnerCorp/build-actions/blob/f3a79ac52d53e8b2218381ae5dd240619ed9eacf/scala/action.yml#L129-L134

**Platform:**
- [x] Ubuntu
- [ ] macOS
- [ ] Windows

**Runner type:**
- [x] Hosted
- [ ] Self-hosted

**Repro steps:**  
1. Create a workflow that `runs-on: ubuntu-latest` or `runs-on: ubuntu-24.04` 
2. Add a step that either `uses: actions/setup-java@v4` or `uses: GarnerCorp/build-actions/scala@0cc885d8117f6bb9c89dc99a34f5edebeac115cf`
3. If you aren't using the build-actions/scala action which automatically uses `sbt`, run `sbt`

**Expected behavior:**
`sbt` should work

**Actual behavior:**
`sbt: not found`

**Comment by gowridurgad** [comment]
Hello @jsoref,
Thank you for creating this issue. We will investigate it and provide feedback as soon as we have some updates.

**Comment by boydsa92** [comment]
We are facing the same issue
`line 2: sbt: command not found`

**Comment by mahabaleshwars** [comment]
Hi @jsoref,

You're absolutely right that sbt is currently not included in Ubuntu 24.04. The Runner Image team has made cuts to the list of included packages in order to maintain our SLA for free disk space. More details here: https://github.com/actions/runner-images/issues/10636.

Since sbt is not part of the runner image, CI workflows were failing, which is why the workflows have been updated.

The setup-java action provides the environment for sbt, and including sbt by default depends on the runner image configuration.

You can still install sbt manually using ``` sudo apt-get install sbt ```, but currently, it isn't working with Ubuntu 24.04. Below is a workaround for the issue.

 ``` name: Install sbt
      run: |
        echo ""deb https://repo.scala-sbt.org/scalasbt/debian all main"" | sudo tee /etc/apt/sources.list.d/sbt.list
        echo ""deb https://repo.scala-sbt.org/scalasbt/debian /"" | sudo tee /etc/apt/sources.list.d/sbt_old.list
        curl -sL ""https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823"" | sudo apt-key add
        sudo apt-get update
        sudo apt-get install sbt
```
 Let me know if it helps resolve your problem.

**Comment by jsoref** [comment]
first, your sample code is buggy and you really shouldn't suggest it -- the `""` evaluate to empty.
second, [`apt-key` is deprecated](https://medium.com/@glegoux/ubuntu-22-04-jammy-jellyfish-apt-key-is-deprecated-2dbbee8aec84) and has been since at least ubuntu 22.04 -- suggesting it anywhere is a horrible idea.

third, this action is designed to support sbt -- each consumer of this action shouldn't have to try to avoid making the mistakes that you included in your sample code (or the various other mistakes that may be present or properly maintain it for when the key for ubuntu changes). Or the various other problems entailed here.

**Comment by SethTisue** [comment]
The maintainers of sbt have made this action available: https://github.com/sbt/setup-sbt  cc @eed3si9n

So all you need to do is add `uses: sbt/setup-sbt@v1` and you're good again

**Comment by mahabaleshwars** [comment]
Hi @jsoref ,

The double quotes "" I used in the original sample code were indeed incorrect, as they would evaluate to empty strings. I appreciate you pointing that out. Here’s the corrected command:

```
run:|
      sudo apt-get update
      sudo apt-get install apt-transport-https curl gnupg -yqq
      echo ""deb https://repo.scala-sbt.org/scalasbt/debian all main"" | sudo tee /etc/apt/sources.list.d/sbt.list
      echo ""deb https://repo.scala-sbt.org/scalasbt/debian /"" | sudo tee /etc/apt/sources.list.d/sbt_old.list
      curl -sL ""https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823"" | sudo -H gpg --no-default-keyring --keyring gnupg-ring:/etc/apt/trusted.gpg.d/scalasbt-release.gpg --import
      sudo chmod 644 /etc/apt/trusted.gpg.d/scalasbt-release.gpg
      sudo apt-get update
      sudo apt-get install sbt
```
[Reference Link](https://www.scala-sbt.org/1.x/docs/Installing-sbt-on-Linux.html#Ubuntu+and+other+Debian-based+distributions)

You're right that the setup-java action is designed to support sbt out of the box, and consumers shouldn't have to manually manage dependencies like sbt for typical use cases. The issue here is that sbt is not included in the default Ubuntu 24.04 runner image, and because of this, CI workflows are failing. The Runner Image team made this change to reduce disk usage. The workaround I provided was meant to help in cases where users are facing this problem, but I do agree it’s not an ideal long-term solution.

**Comment by jsoref** [comment]
It doesn't look like you've fixed the doubled quotes. You did fix the apt-key things, which is nice although that's not an alternative command I've seen before...

**Comment by SethTisue** [comment]
I agree it would make a lot of sense for setup-java to pick this up now that it isn't in the ubuntu-latest anymore.

**Comment by eed3si9n** [comment]
I recommend using [setup-sbt](https://github.com/sbt/setup-sbt), or if setup-java folks are considering to make `sbt` runner script available, take the similar approach that I took and use the universal `.zip` or `.tgz` file instead of `apt-get install sbt`:
1. Deb / RPM packages are hosted by an Artifactory instance, which Scala Center is granted usage by JFrog for free, renewed annually, and we should operate under the assumption that it might someday go away. Thus, in general we should avoid using `apt-get install sbt`.
2. For the reason above, I have reduced the file size of `.deb` file by not including some files `.zip` contains (so it would be cheaper for whoever who might have to run the Artifactory instance), and the `sbt` runner script would download them on the fly.
3. Deb package works only for Linux, but GitHub Action supports macOS and Windows as well.

**Comment by mahabaleshwars** [comment]
Hello @jsoref,

I’ve updated the comment by removing the extra double quotes. Please let us know if the feedback resolves the issue.

Currently, setup-java supports the features listed here: [setup-java documentation](https://github.com/actions/setup-java?tab=readme-ov-file#setup-java), but sbt is not included by default. Tools need to be installed on the runner image or can be added via workarounds. You can raise an issue here to request sbt on the ubuntu-latest runner (Ubuntu 24.04): [runner-images issues.](https://github.com/actions/runner-images/issues)

**Comment by jsoref** [comment]
@mahabaleshwars: If I owned this repository (and I wasn't willing to just fix setup-java to install sbt), I'd change the readme:

```
-Caching dependencies managed by sbt.
+Caching dependencies managed by sbt (sbt is not necessarily installed by default, ...)
```

the `...` would depend on how I was feeling, I might just link to this ticket since it provides two possible fixes (https://github.com/actions/setup-java/issues/712#issuecomment-2563623521 and https://github.com/actions/setup-java/issues/712#issuecomment-2564193070). Or I might suggest one or the other. Offhand, if you aren't particularly interested in maintaining any support for sbt, I'd be more inclined to point to the setup-sbt action than the code you've listed, but it's just a slight preference.

I'm not going to file a ticket asking the runners team to change their images, that's pointless -- but thank you for the invitation.

**Comment by mahabaleshwars** [comment]
Hi @jsoref,

Please refer to this issue:  [#11350](https://github.com/actions/runner-images/issues/11350). You can follow that for updates regarding sbt on Ubuntu Latest (Ubuntu 24.04).

As the maintainers of setup-java, we can support sbt, but we are unable to provide the tools by default. This falls under the responsibility of the runner images team, and they would need to include the necessary tools.

I'm going to close this issue now. Please feel free to follow the linked issue for any progress or updates.
